File: src/main/java/com/gestiontache/service/TaskService.java
package com.gestiontache.service;

import com.gestiontache.model.Recurrence;
import com.gestiontache.model.Task;
import com.gestiontache.repository.TaskRepository;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * In-memory task list backed by {@link TaskRepository}, which stores one
 * JSON file per day. Every mutation is persisted immediately, rewriting only
 * the day file(s) actually affected (never the whole dataset), so the local
 * files always reflect the current state.
 */
public class TaskService {

    private final TaskRepository repository;
    private final List<Task> tasks;

    public TaskService(TaskRepository repository) {
        this.repository = repository;
        this.tasks = repository.loadAll();
    }

    public List<Task> getTasksForDate(LocalDate date) {
        return tasks.stream()
                .filter(t -> date.equals(t.getDate()))
                .sorted(Comparator.comparingInt(Task::getOrder).thenComparing(Task::getCreatedAt))
                .collect(Collectors.toList());
    }

    /** Searches every task, across every day file, by title or description. */
    public List<Task> search(String keyword) {
        return tasks.stream()
                .filter(t -> t.matches(keyword))
                .sorted(Comparator.comparing(Task::getDate).reversed())
                .collect(Collectors.toList());
    }

    public List<Task> getOverdueUnfinishedTasks(LocalDate today) {
        return tasks.stream()
                .filter(t -> !t.isCompleted() && t.getDate().isBefore(today))
                .collect(Collectors.toList());
    }

    public void addTask(Task task) {
        task.setOrder(nextOrderForDate(task.getDate()));
        tasks.add(task);
        persistDate(task.getDate());
    }

    /**
     * Applies a new manual order to the tasks of {@code date}, following the
     * order given by {@code orderedTasks} (typically the list view's items
     * after a drag-and-drop move).
     */
    public void reorderTasksForDate(LocalDate date, List<Task> orderedTasks) {
        int order = 0;
        for (Task t : orderedTasks) {
            if (date.equals(t.getDate())) {
                t.setOrder(order++);
            }
        }
        persistDate(date);
    }

    /**
     * The task object is mutated in place by the caller (including its date,
     * for an edit that moves it to another day); this persists whichever day
     * file(s) are actually affected.
     */
    public void updateTask(Task task, LocalDate previousDate) {
        if (previousDate.equals(task.getDate())) {
            persistDate(task.getDate());
        } else {
            persistDates(Set.of(previousDate, task.getDate()));
        }
    }

    public void deleteTask(Task task) {
        LocalDate date = task.getDate();
        tasks.remove(task);
        persistDate(date);
    }

    /**
     * Marks a task done or not. When a task with a recurrence rule
     * transitions to completed, the next occurrence is created automatically
     * (not completed) on the corresponding future date.
     */
    public void setCompleted(Task task, boolean completed) {
        boolean wasCompleted = task.isCompleted();
        task.setCompleted(completed);
        if (completed && !wasCompleted && task.getRecurrence() != Recurrence.AUCUNE) {
            Task next = createNextOccurrence(task);
            persistDates(Set.of(task.getDate(), next.getDate()));
        } else {
            persistDate(task.getDate());
        }
    }

    private Task createNextOccurrence(Task task) {
        LocalDate nextDate = task.getRecurrence().nextOccurrence(task.getDate());
        Task next = new Task(task.getTitle(), task.getDescription(), nextDate);
        next.setPriority(task.getPriority());
        next.setRecurrence(task.getRecurrence());
        next.setOrder(nextOrderForDate(nextDate));
        tasks.add(next);
        return next;
    }

    /**
     * Reports every unfinished task of {@code date} to the next day.
     * @return the number of tasks moved.
     */
    public int reportUnfinishedToNextDay(LocalDate date) {
        List<Task> unfinished = tasks.stream()
                .filter(t -> date.equals(t.getDate()) && !t.isCompleted())
                .collect(Collectors.toList());
        LocalDate next = date.plusDays(1);
        int order = nextOrderForDate(next);
        for (Task t : unfinished) {
            t.setDate(next);
            t.setOrder(order++);
        }
        if (!unfinished.isEmpty()) {
            persistDates(Set.of(date, next));
        }
        return unfinished.size();
    }

    /**
     * Reports every unfinished task from any previous day to {@code today}.
     * @return the number of tasks moved.
     */
    public int reportOverdueToToday(LocalDate today) {
        List<Task> overdue = getOverdueUnfinishedTasks(today);
        Set<LocalDate> affectedDates = new HashSet<>();
        for (Task t : overdue) {
            affectedDates.add(t.getDate());
        }
        affectedDates.add(today);

        int order = nextOrderForDate(today);
        for (Task t : overdue) {
            t.setDate(today);
            t.setOrder(order++);
        }
        if (!overdue.isEmpty()) {
            persistDates(affectedDates);
        }
        return overdue.size();
    }

    private int nextOrderForDate(LocalDate date) {
        return tasks.stream()
                .filter(t -> date.equals(t.getDate()))
                .mapToInt(Task::getOrder)
                .max()
                .orElse(-1) + 1;
    }

    /** Rewrites the day file for {@code date} from the current in-memory state. */
    private void persistDate(LocalDate date) {
        List<Task> forDate = tasks.stream()
                .filter(t -> date.equals(t.getDate()))
                .collect(Collectors.toList());
        repository.saveDay(date, forDate);
    }

    private void persistDates(Set<LocalDate> dates) {
        for (LocalDate date : dates) {
            persistDate(date);
        }
    }
}

File: dotnet-winui3/GestionTaches/Services/TaskService.cs
using GestionTaches.Models;

namespace GestionTaches.Services;

/// <summary>
/// In-memory task list backed by <see cref="TaskRepository"/>. Every
/// mutation is persisted immediately, rewriting only the day file(s)
/// actually affected. Mirrors the JavaFX client's TaskService (core MVP
/// subset: no recurrence/subtasks/attachments/statistics here).
/// </summary>
public class TaskService
{
    private readonly TaskRepository _repository;
    private readonly List<TaskItem> _tasks;

    public TaskService(TaskRepository repository)
    {
        _repository = repository;
        _tasks = repository.LoadAll();
    }

    public List<TaskItem> GetTasksForDate(DateOnly date) =>
        _tasks.Where(t => t.Date == date)
            .OrderBy(t => t.Order)
            .ThenBy(t => t.CreatedAt)
            .ToList();

    /// <summary>Searches every task, across every day file, by title or description.</summary>
    public List<TaskItem> Search(string? keyword) =>
        _tasks.Where(t => t.Matches(keyword))
            .OrderByDescending(t => t.Date)
            .ToList();

    public List<TaskItem> GetOverdueUnfinishedTasks(DateOnly today) =>
        _tasks.Where(t => !t.Completed && t.Date < today).ToList();

    public void AddTask(TaskItem task)
    {
        task.Order = NextOrderForDate(task.Date);
        _tasks.Add(task);
        PersistDate(task.Date);
    }

    /// <summary>
    /// The task object is mutated in place by the caller (including its
    /// date, for an edit that moves it to another day); this persists
    /// whichever day file(s) are actually affected.
    /// </summary>
    public void UpdateTask(TaskItem task, DateOnly previousDate)
    {
        if (previousDate == task.Date)
        {
            PersistDate(task.Date);
        }
        else
        {
            PersistDates(new HashSet<DateOnly> { previousDate, task.Date });
        }
    }

    public void DeleteTask(TaskItem task)
    {
        DateOnly date = task.Date;
        _tasks.Remove(task);
        PersistDate(date);
    }

    public void SetCompleted(TaskItem task, bool completed)
    {
        task.Completed = completed;
        PersistDate(task.Date);
    }

    /// <summary>Reports every unfinished task of <paramref name="date"/> to the next day. Returns the number of tasks moved.</summary>
    public int ReportUnfinishedToNextDay(DateOnly date)
    {
        List<TaskItem> unfinished = _tasks.Where(t => t.Date == date && !t.Completed).ToList();
        DateOnly next = date.AddDays(1);
        int order = NextOrderForDate(next);
        foreach (TaskItem t in unfinished)
        {
            t.Date = next;
            t.Order = order++;
        }

        if (unfinished.Count > 0)
        {
            PersistDates(new HashSet<DateOnly> { date, next });
        }

        return unfinished.Count;
    }

    /// <summary>Reports every unfinished task from any previous day to <paramref name="today"/>. Returns the number of tasks moved.</summary>
    public int ReportOverdueToToday(DateOnly today)
    {
        List<TaskItem> overdue = GetOverdueUnfinishedTasks(today);
        var affectedDates = new HashSet<DateOnly>();
        foreach (TaskItem t in overdue)
        {
            affectedDates.Add(t.Date);
        }
        affectedDates.Add(today);

        int order = NextOrderForDate(today);
        foreach (TaskItem t in overdue)
        {
            t.Date = today;
            t.Order = order++;
        }

        if (overdue.Count > 0)
        {
            PersistDates(affectedDates);
        }

        return overdue.Count;
    }

    private int NextOrderForDate(DateOnly date)
    {
        var ordersForDate = _tasks.Where(t => t.Date == date).Select(t => t.Order).ToList();
        return (ordersForDate.Count > 0 ? ordersForDate.Max() : -1) + 1;
    }

    /// <summary>Rewrites the day file for <paramref name="date"/> from the current in-memory state.</summary>
    private void PersistDate(DateOnly date)
    {
        List<TaskItem> forDate = _tasks.Where(t => t.Date == date).ToList();
        _repository.SaveDay(date, forDate);
    }

    private void PersistDates(IEnumerable<DateOnly> dates)
    {
        foreach (DateOnly date in dates)
        {
            PersistDate(date);
        }
    }
}
